# Mahalanobis Distance

Mahalanobis Distance is a multivariate measure of the distance between a point and a distribution. It's an effective way to detect outliers, especially when dealing with multidimensional data.

#### Steps to Calculate Mahalanobis Distance:

1. Calculate the covariance matrix of the dataset.
2. Calculate the inverse of the covariance matrix.
3. For each data point (x), calculate its Mahalanobis Distance using the formula: D² = (x - μ)ᵀ * S⁻¹ * (x - μ), where S⁻¹ is the inverse covariance matrix.

#### Mahalanobis Distance Formula

$$ D^2 = (x - \mu)^T S^{-1} (x - \mu) $$

where:
- $x$ is the data point.
- $\mu$ is the mean of the dataset.
- $S$ is the covariance matrix of the dataset.
- $S^{-1}$ is the inverse of the covariance matrix.

## Test Platform : Intel Mac on Dev Container (miniconda)

In [ ]:
from sklearn.neighbors import LocalOutlierFactor
import numpy as np
# Sample data
X = [[10, 10], [12, 12], [11, 11], [14, 14], [100, 100], [14, 14]]
clf = LocalOutlierFactor(n_neighbors=2)
outlier_labels = clf.fit_predict(X)
# -1 indicates an outlier
print(f"Outlier labels: {outlier_labels}")

: 

In [28]:
# Outlier detection using Mahalanobis distance
from scipy.stats import chi2
X = np.array([[10, 10], [12, 12], [11, 11], [14, 14], [100, 100], [14, 14]])
# Calculate mean and covariance matrix

mean = np.mean(X, axis=0)
cov_matrix = np.cov(X.T)
inv_cov_matrix = np.linalg.inv(cov_matrix)
cov_matrix

array([[1287.36666667, 1287.36666667],
       [1287.36666667, 1287.36666667]])

In [29]:
inv_cov_matrix

array([[ 4.39804651e+12, -4.39804651e+12],
       [-4.39804651e+12,  4.39804651e+12]])

In [30]:
# Calculate Mahalanobis Distance for each data point
mahalanobis_distances = np.array([np.dot(np.dot((x - mean), inv_cov_matrix), (x - mean).T) for x in X])
mahalanobis_distances

array([0.13835992, 0.10743544, 0.1224094 , 0.08041721, 2.61394586,
       0.08041721])

In [31]:
# Compare with a critical value from the chi-squared distribution
critical_value = chi2.ppf((1-0.1), df=2)  # df is the number of dimensions
outliers = np.where(mahalanobis_distances > critical_value)
critical_value

4.605170185988092

In [32]:
print(f"Outlier indices: {outliers[0]}")

Outlier indices: []


**The critical value is 4.605 whereas distance is 2.614**

## Test Platform : Intel Mac on uv

In [2]:
from sklearn.neighbors import LocalOutlierFactor
import numpy as np
# Sample data
X = [[10, 10], [12, 12], [11, 11], [14, 14], [100, 100], [14, 14]]
clf = LocalOutlierFactor(n_neighbors=2)
outlier_labels = clf.fit_predict(X)
# -1 indicates an outlier
print(f"Outlier labels: {outlier_labels}")

Outlier labels: [ 1  1  1  1 -1  1]


In [3]:
# Outlier detection using Mahalanobis distance
from scipy.stats import chi2
X = np.array([[10, 10], [12, 12], [11, 11], [14, 14], [100, 100], [14, 14]])
# Calculate mean and covariance matrix

mean = np.mean(X, axis=0)
cov_matrix = np.cov(X.T)
inv_cov_matrix = np.linalg.inv(cov_matrix)
cov_matrix

array([[1287.36666667, 1287.36666667],
       [1287.36666667, 1287.36666667]])

In [4]:
inv_cov_matrix

array([[ 4.39804651e+12, -4.39804651e+12],
       [-4.39804651e+12,  4.39804651e+12]])

In [5]:
# Calculate Mahalanobis Distance for each data point
mahalanobis_distances = np.array([np.dot(np.dot((x - mean), inv_cov_matrix), (x - mean).T) for x in X])
mahalanobis_distances

array([0.26302083, 0.11588542, 0.12369792, 0.10026042, 4.57291667,
       0.10026042])

In [6]:
# Compare with a critical value from the chi-squared distribution
critical_value = chi2.ppf((1-0.1), df=2)  # df is the number of dimensions
outliers = np.where(mahalanobis_distances > critical_value)
critical_value

np.float64(4.605170185988092)

In [7]:
print(f"Outlier indices: {outliers[0]}")

Outlier indices: []


**The critical value is 4.605 whereas distance is 4.5729**